### PR TITLE
First Pull

### PR DESCRIPTION
**Pull request by 0xMoJo7**
Everything functioning as expected.
What needs fixin'?

### CODE REVIEW

File: code.py [hunk 1]
```diff
@@ -1,14 +1,26 @@
-c = 5.0
-_mv_db = {"Xmen 8: The Xmennening": 10, "The Bromance" : 20, "Gigli: The Play: The Book: The movie" : 102}
-_mvg_d = {"Bob" : {"movies":[], "cash":100.0}, "Jim" : {"movies": ["Xmen 8: The Xmennening"], "cash":10.0}, "Cary" : {"movies": ["Gigli: The Play: The Book: The movie", "The Bromance"], "cash":120.0}, "Ricci": {"movies": [], "cash":4.0}}
-def do_p(mv, mvg):
-    _mv_db[mv] = _mv_db[mv] - 1
-    if _mvg_d[mvg]["cash"] > 5.0:
+MOVIE_PRICE = 5.0
```
Comment (CoitThomas) [inline code.py]: Move this to movies.py

File: code.py [hunk 2]
```diff
@@ -1,14 +1,26 @@
...
 
+def enough_money(customer):
```
Comment (CoitThomas) [inline code.py]: move this to movies.py as well

File: code.py [hunk 3]
```diff
@@ -1,14 +1,26 @@
...
+        customer[mvg]["cash"] = customer[mvg]["cash"] - 5.0
+        customer[mvg]["movies"].append(mv);
+        return True
 
-        _mvg_d[mvg]["cash"] = _mvg_d[mvg]["cash"] - 5.0; _mvg_d[mvg]["movies"].append(mv); return True
-    return False
-def tesst():
-    do_p("Xmen 8: The Xmennening", "Ricci")
-    assert _mv_db["Xmen 8: The Xmennening"] == 10 #WTF? 
+if __name__ == "__main__":
```
Comment (CoitThomas) [inline code.py]: Nix this and subsequently the file - all of the Customer info and movie info would usually live in a database not in your code. Just use those dictionaries for testing purposes.

File: movies.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
+"""Movie ticket process system and database"""
+
+def enough_money(customer):
+    """Check if customer has adequate funds"""
+    balance = customer["cash"]
+    if balance < 5.0:
```
Comment (CoitThomas) [inline movies.py]: Once you move `MOVIE_PRICE = 5.0` into here, you can start referencing `MOVIE_PRICE` instead of hard coding `5.0` - this way, in the future, if the price of a movie changes, you only have to make 1 change instead of several.

Comment (CoitThomas) [reply movies.py]: It also makes things less "magical" - you know exactly what the 5.0 represents

File: movies.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+def purchase_ticket(movie, customer):
+    """Performs the steps required in purchasing a ticket and returns altered customer and movie"""
+    customer = charge(customer)
+    customer = dispense_ticket(movie, customer)
+    movie = remove_seat(movie)
+    return customer, movie
+
+def charge(customer):
+    """Charge customer $5"""
+    customer["cash"] = customer["cash"] - 5.0
```
Comment (CoitThomas) [inline movies.py]: MOVIE_PRICE

File: movies.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+def purchase_ticket(movie, customer):
+    """Performs the steps required in purchasing a ticket and returns altered customer and movie"""
+    customer = charge(customer)
+    customer = dispense_ticket(movie, customer)
+    movie = remove_seat(movie)
+    return customer, movie
+
+def charge(customer):
+    """Charge customer $5"""
```
Comment (CoitThomas) [inline movies.py]: """Charge the customer the price of 1 movie ticket"""

Comment (CoitThomas) [reply movies.py]: Be nice to your future self - that movie ticket price is probably going to change - you don't want future Eric having to hunt down outdated comments

File: movies.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    return customer
+
+def dispense_ticket(movie, customer):
+    """Add movie to customer's account"""
+    customer["movies"].append(movie["title"])
+    return customer
+
+def remove_seat(movie):
+    """Remove seat from ticket database"""
+    movie["seats"] = movie["seats"] - 1.0
```
Comment (CoitThomas) [inline movies.py]: Shouldn't this be an integer? You can't have fractions of a movie seat.

File: movies.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    return customer, movie
+
+def add_funds(customer):
+    """Add $5 back to moviegoer's account"""
+    customer["cash"] = customer["cash"] + 5.0
+    return customer
+
+def remove_ticket(movie, customer):
+    """Removies ticket from customers library"""
+    customer["movies"].remove(movie["title"])
```
Comment (CoitThomas) [inline movies.py]: This looks like a great opportunity for an assertion. Assert that the customer actually has that movie you want to remove in their library before removing it. Go back through your other methods and be sure to utilize assertions in them as well.

File: movies.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    return movie
+
+def print_movie_data(movie):
+    """Prints movie database informations"""
+    print("Movie Title:           ", movie["title"])
+    print("Seats Remaining:       ", movie["seats"])
+
+def print_customer_data(customer):
+    """Prints customer database information"""
+    print("Movie Tickets:         ", customer["movies"])
```
Comment (CoitThomas) [inline movies.py]: The output of this will give you the movies in brackets: `[MovieA, MovieB, MovieC]` - try to find a cleaner way to present this.

File: movies.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    """Removies ticket from customers library"""
+    customer["movies"].remove(movie["title"])
+    return customer
+
+def add_seat(movie):
+    """Add seat back to database"""
+    movie["seats"] = movie["seats"] + 1.0
+    return movie
+
+def print_movie_data(movie):
```
Comment (CoitThomas) [inline movies.py]: Add another print function that you can just feed an entire dictionary of movies to. You'll be able to utilize this function within the new one.

File: movies.py [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    """Prints movie database informations"""
+    print("Movie Title:           ", movie["title"])
+    print("Seats Remaining:       ", movie["seats"])
+
+def print_customer_data(customer):
```
Comment (CoitThomas) [inline movies.py]: Do the same thing here. Create another print function that can be fed an entire dictionary of customers.

File: movies_test.py [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+              "cash" : 120.0},
+    "Ricci": {"movies": [],
+              "cash" : 4.0}
+}
+MOVIE_TICKETS = {
+    "Xmen" : {"title" : "Xmen 8: The Xmennening", "seats" : 10},
+    "Bromance" : {"title" : "The Bromance", "seats" : 20},
+    "Gigli" : {"title" : "Gigli: The Play: The Book: The movie", "seats" : 102}
+    }
+def test_enough_money():
```
Comment (CoitThomas) [inline movies_test.py]: New line

File: movies_test.py [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+    """Tests charge function"""
+    bob = {"movies": [], "cash": 100.0}
+    bobs_actual_cash = movies.charge(bob)["cash"]
+    bobs_expected_cash = 95.0
+    assert bobs_actual_cash == bobs_expected_cash
+    jim = {"movies": "Xmen 8: The Xmennening", "cash" : 10.0}
+    jims_actual_cash = movies.charge(jim)["cash"]
+    jims_expected_cash = 5.0
+    assert jims_actual_cash == jims_expected_cash
+def test_dispense_ticket():
```
Comment (CoitThomas) [inline movies_test.py]: New line

File: movies_test.py [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+    xmen = {"title" : "Xmen 8: The Xmennening", "seats" : 10}
+    bobs_library = movies.dispense_ticket(xmen, bob)["movies"]
+    bobs_expected_library = ["Xmen 8: The Xmennening"]
+    assert bobs_library == bobs_expected_library
+    jim = {"movies": ["Xmen 8: The Xmennening"], "cash" : 10.0}
+    bromance = {"title" : "The Bromance", "seats" : 20}
+    jims_library = movies.dispense_ticket(bromance, jim)["movies"]
+    jims_expected_library = ["Xmen 8: The Xmennening", "The Bromance"]
+    assert jims_library == jims_expected_library
+def test_remove_seat():
```
Comment (CoitThomas) [inline movies_test.py]: New line

File: movies_test.py [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+def test_purchase_tickets():
+    """Tests all subfunctions of purchase_tickets"""
+    bob = {"movies": [], "cash": 100.0}
+    bromance = {"title" : "The Bromance", "seats" : 20}
+    actual_bob, actual_bromance = movies.purchase_ticket(bromance, bob)
+    expected_bob = {"movies": ["The Bromance"], "cash" : 95.0}
+    expected_bromance = {"title" : "The Bromance", "seats" : 19}
+    assert actual_bob == expected_bob
+    assert actual_bromance == expected_bromance
+def test_add_funds():
```
Comment (CoitThomas) [inline movies_test.py]: New line

File: movies_test.py [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+    """Tests charge function"""
+    bob = {"movies": [], "cash": 100.0}
+    bobs_actual_cash = movies.add_funds(bob)["cash"]
+    bobs_expected_cash = 105.0
+    assert bobs_actual_cash == bobs_expected_cash
+    ricci = {"movies" : [], "cash" : 4.0}
+    ricci_actual_cash = movies.add_funds(ricci)["cash"]
+    ricci_expected_cash = 9.0
+    assert ricci_actual_cash == ricci_expected_cash
+def test_remove_ticket():
```
Comment (CoitThomas) [inline movies_test.py]: New line

File: movies_test.py [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+    jims_library = movies.remove_ticket(xmen, jim)["movies"]
+    jims_expected_library = []
+    assert jims_library == jims_expected_library
+    cary = {"movies": ["Gigli: The Play: The Book: The movie", "The Bromance"],
+            "cash" : 120.0}
+    bromance = {"title" : "The Bromance", "seats" : 20}
+    cary_library = movies.remove_ticket(bromance, cary)["movies"]
+    cary_expected_library = ["Gigli: The Play: The Book: The movie"]
+    assert cary_library == cary_expected_library
+def test_add_seat():
```
Comment (CoitThomas) [inline movies_test.py]: New line

File: movies_test.py [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+              "cash" : 4.0}
+}
+MOVIE_TICKETS = {
+    "Xmen" : {"title" : "Xmen 8: The Xmennening", "seats" : 10},
+    "Bromance" : {"title" : "The Bromance", "seats" : 20},
+    "Gigli" : {"title" : "Gigli: The Play: The Book: The movie", "seats" : 102}
+    }
+def test_enough_money():
+    """Tests the enough money function"""
+    bob = {"movies":[], "cash" : 100.0}
```
Comment (CoitThomas) [inline movies_test.py]: Instead of building a new customer, just use the ones that you already made in `CUSTOMERS`.
For example: `bob = CUSTOMERS["Bob"]`

File: movies_test.py [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+    bobs_expected_cash = 95.0
+    assert bobs_actual_cash == bobs_expected_cash
+    jim = {"movies": "Xmen 8: The Xmennening", "cash" : 10.0}
+    jims_actual_cash = movies.charge(jim)["cash"]
+    jims_expected_cash = 5.0
+    assert jims_actual_cash == jims_expected_cash
+def test_dispense_ticket():
+    """Test adding ticket to customers account"""
+    bob = {"movies": [], "cash": 100.0}
+    xmen = {"title" : "Xmen 8: The Xmennening", "seats" : 10}
```
Comment (CoitThomas) [inline movies_test.py]: Same thing with movies: `xmen = MOVIES["Xmen"]`

File: movies_test.py [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+    "Bob" : {"movies":[],
+             "cash" : 100.0},
+    "Jim" : {"movies": ["Xmen 8: The Xmennening"],
+             "cash" : 10.0},
+    "Cary" : {"movies": ["Gigli: The Play: The Book: The movie", "The Bromance"],
+              "cash" : 120.0},
+    "Ricci": {"movies": [],
+              "cash" : 4.0}
+}
+MOVIE_TICKETS = {
```
Comment (CoitThomas) [inline movies_test.py]: I think I told you to call this MOVIE_TICKETS at one point, but after seeing it again, I think MOVIES is probably a better choice

File: movies_test.py [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+    "Xmen" : {"title" : "Xmen 8: The Xmennening", "seats" : 10},
```
Comment (CoitThomas) [inline movies_test.py]: I would make "seats" a bit more descriptive. Something like "seats_available" or "tickets_remaining".

File: movies_test.py [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+    "Bromance" : {"title" : "The Bromance", "seats" : 20},
+    "Gigli" : {"title" : "Gigli: The Play: The Book: The movie", "seats" : 102}
+    }
+def test_enough_money():
+    """Tests the enough money function"""
+    bob = {"movies":[], "cash" : 100.0}
+    ricci = {"movies": [], "cash" : 4.0}
+    assert movies.enough_money(ricci) is False
```
Comment (CoitThomas) [inline movies_test.py]: Assertions of booleans do not require `is bool`. You can just do:
`assert not movies.enough_money(ricci)`
`assert movies.enough_money(bob)`

File: movies_test.py [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+    bro_seats = movies.remove_seat(bromance)
+    bro_expected_seats = {"title" : "The Bromance", "seats" : 19}
+    assert bro_seats == bro_expected_seats
+
+def test_purchase_tickets():
+    """Tests all subfunctions of purchase_tickets"""
+    bob = {"movies": [], "cash": 100.0}
+    bromance = {"title" : "The Bromance", "seats" : 20}
+    actual_bob, actual_bromance = movies.purchase_ticket(bromance, bob)
+    expected_bob = {"movies": ["The Bromance"], "cash" : 95.0}
```
Comment (CoitThomas) [inline movies_test.py]: Hard coding these numbers will be a big pain in the ass later. Try something along the lines of:
`expected_bob = {"movies": ["The Bromance"], "cash" : bob["cash"] - MOVIE_PRICE}`

**Review by CoitThomas — CHANGES_REQUESTED** [review]
Despite the multiple comments, this is a good start. I will probably have more after you make these changes, but I figure this is enough for now. Let me know if you have any questions.

File: movies.py [hunk 24]
```diff
@@ -0,0 +1,100 @@
+"""Movie ticket process system and database"""
+
+import pprint
+
+MOVIE_PRICE = 5.0
+
+def customer_db():
```
Comment (CoitThomas) [inline movies.py]: This and the movie_db() function should really only be used in your testing module.

File: movies.py [hunk 25]
```diff
@@ -0,0 +1,100 @@
...
+    """Prints movie database informations"""
+    print("Movie Title:           ", movie["title"])
+    print("Seats Remaining:       ", movie["seats_avaliable"])
+
+def print_customer_data(customer): #movies have brackets around them
+    """Prints customer database information"""
+    print("Movie Tickets:         ", ",  ".join(customer["movies"]))
+    print("Account Balance:       ", customer["cash"])
+
+def print_all_movie_data():
```
Comment (CoitThomas) [inline movies.py]: Right now this function will only print the dictionary you've defined up above. This function and its sister function, print_all_customer_data(), should take in a dictionary of their respective data type. That will give these methods greater usability. You should probably also have some asserts in these functions to make sure they are the right type of dictionaries.

File: movies.py [hunk 26]
```diff
@@ -0,0 +1,100 @@
...
+    """Performs the steps required in purchasing a ticket and returns altered customer and movie"""
+    assert enough_money(customer)
+    customer = charge(customer)
+    customer = dispense_ticket(movie, customer)
+    movie = remove_seat(movie)
+    return customer, movie
+
+def charge(customer):
+    """Charge customer the price of one ticket"""
+    customer["cash"] = customer["cash"] - MOVIE_PRICE
```
Comment (CoitThomas) [inline movies.py]: I know it might seem redundant, but having functions which are as self-sufficient as possible can be pretty important, so I would assert that the customer has enough money in this method also. That way, let's say later you write a new feature that calls the charge() method,  you don't accidentally send people into debt. I would also add another test that captures this scenario.

File: movies_test.py [hunk 27]
```diff
@@ -0,0 +1,134 @@
...
+
+def test_enough_money():
+    """Tests the enough money function"""
+    customers = customer_db()
+    bob = customers["Bob"]
+    ricci = customers["Ricci"]
+    assert not movies.enough_money(ricci)
+    assert movies.enough_money(bob)
+
+def test_charge():
```
Comment (CoitThomas) [inline movies_test.py]: Just reiterating, there should be an assertion in here to make sure you aren't sending people into debt.

File: movies_test.py [hunk 28]
```diff
@@ -0,0 +1,134 @@
...
+    bobs_library = movies.dispense_ticket(xmen, bob)["movies"]
+    bobs_expected_library = ["Xmen 8: The Xmennening"]
+    assert bobs_library == bobs_expected_library
+    jim = customers["Jim"]
+    bromance = movies_["Bromance"]
+    jims_library = movies.dispense_ticket(bromance, jim)["movies"]
+    jims_expected_library = ["Xmen 8: The Xmennening", "The Bromance"]
+    assert jims_library == jims_expected_library
+
+def test_remove_seat():
```
Comment (CoitThomas) [inline movies_test.py]: there should also be an assertion in here that makes sure you aren't removing seats which don't exist

File: movies.py [hunk 29]
```diff
@@ -0,0 +1,82 @@
+"""Movie ticket process system and database"""
```
Comment (CoitThomas) [inline movies.py]: Whoops, I missed this as well - no db stuff is in here

File: movies.py [hunk 30]
```diff
@@ -0,0 +1,82 @@
...
+    balance = customer["cash"]
+    if balance < MOVIE_PRICE:
+        return False
+    return True
+
+def purchase_ticket(movie, customer):
+    """Performs the steps required in purchasing a ticket and returns altered customer and movie"""
+    assert enough_money(customer)
+    customer = charge(customer)
+    customer = dispense_ticket(movie, customer)
```
Comment (CoitThomas) [inline movies.py]: you can chain these two function calls on line 15 and 16:
`customer = dispense_ticket(movie, charge(customer))`

File: movies.py [hunk 31]
```diff
@@ -0,0 +1,82 @@
...
+    movie = remove_seat(movie)
+    return customer, movie
+
+def charge(customer):
+    """Charge customer the price of one ticket"""
+    assert enough_money(customer)
+    customer["cash"] = customer["cash"] - MOVIE_PRICE
+    return customer
```
Comment (CoitThomas) [inline movies.py]: let's also condense down line 23 and 24 to just:
`return customer["cash"] - MOVIE_PRICE`

File: movies.py [hunk 32]
```diff
@@ -0,0 +1,82 @@
...
+
+def dispense_ticket(movie, customer):
+    """Add movie to customer's account"""
+    assert movie["seats_avaliable"] >= 1
+    customer["movies"].append(movie["title"])
+    return customer
```
Comment (CoitThomas) [inline movies.py]: same here:
`return customer["movies"].append(movie["title"])`

File: movies.py [hunk 33]
```diff
@@ -0,0 +1,82 @@
...
+
+def remove_seat(movie):
+    """Remove seat from ticket database"""
+    assert movie["seats_avaliable"] >= 1
+    movie["seats_avaliable"] = movie["seats_avaliable"] - 1
+    return movie
```
Comment (CoitThomas) [inline movies.py]: condense

File: movies.py [hunk 34]
```diff
@@ -0,0 +1,82 @@
...
+    """Performs the steps required in refunding a ticket and returns altered customer and movie"""
+    customer = add_funds(customer)
+    customer = remove_ticket(movie, customer)
+    movie = add_seat(movie)
+    return customer, movie
+
+def add_funds(customer):
+    """Add the value of one ticket to moviegoer's account"""
+    customer["cash"] = customer["cash"] + MOVIE_PRICE
+    return customer
```
Comment (CoitThomas) [inline movies.py]: condense

File: movies.py [hunk 35]
```diff
@@ -0,0 +1,82 @@
...
+
+def remove_ticket(movie, customer):
+    """Removies ticket from customers library"""
+    assert movie["title"] in customer["movies"]
+    customer["movies"].remove(movie["title"])
+    return customer
```
Comment (CoitThomas) [inline movies.py]: condense

File: movies.py [hunk 36]
```diff
@@ -0,0 +1,82 @@
...
+
+def add_seat(movie):
+    """Add seat back to database"""
+    movie["seats_avaliable"] = movie["seats_avaliable"] + 1
+    return movie
```
Comment (CoitThomas) [inline movies.py]: condense

File: movies_test.py (hunk 28 above)
Comment (CoitThomas) [inline movies_test.py]: So what I was getting at with this one, is that you should have a test in here similar to what you did here:
``` 
+    ricci = customers["Ricci"]
 +    assert not movies.enough_money(ricci)
```
create a move that has zero seats and then try to remove a seat from it

File: movies_test.py [hunk 37]
```diff
@@ -0,0 +1,140 @@
...
+
+def test_charge():
+    """Tests charge function"""
+    customers = customer_db()
+    bob = customers["Bob"]
+    assert movies.enough_money(bob)
+    bobs_actual_cash = movies.charge(bob)["cash"]
+    bobs_expected_cash = 95.0
+    assert bobs_actual_cash == bobs_expected_cash
+    jim = customers["Jim"]
```
Comment (CoitThomas) [inline movies_test.py]: let's get some whitespace in here to separate out the tests within each test - you should go ahead and do this for all of your tests - otherwise, it can be a bit muddled and harder to follow what is actually being tested

File: movies.py (hunk 30 above)
Comment (0xMoJo7) [reply movies.py]: I get why this makes sense, and makes it shorter. But for the sake of readability --  wouldn't two lines be preferred?

Comment (CoitThomas) [reply movies.py]: This improves the readability by making it more concise.
i.e. - "dispense the ticket for the movie and charge the customer"

File: movies.py (hunk 31 above)
Comment (CoitThomas) [inline movies.py]: ignore this - won't work because it's only altering a single field

File: movies.py (hunk 32 above)
Comment (CoitThomas) [inline movies.py]: ignore

File: movies.py [hunk 38]
```diff
@@ -0,0 +1,82 @@
...
+def remove_seat(movie):
+    """Remove seat from ticket database"""
+    assert movie["seats_avaliable"] >= 1
+    movie["seats_avaliable"] = movie["seats_avaliable"] - 1
+    return movie
+
+def refund_ticket(movie, customer):
+    """Performs the steps required in refunding a ticket and returns altered customer and movie"""
+    customer = add_funds(customer)
+    customer = remove_ticket(movie, customer)
```
Comment (CoitThomas) [inline movies.py]: this one actually can be condensed

File: movies.py (hunk 33 above)
Comment (CoitThomas) [inline movies.py]: ignore

File: movies.py (hunk 34 above)
Comment (CoitThomas) [inline movies.py]: ignore

File: movies.py (hunk 35 above)
Comment (CoitThomas) [inline movies.py]: ignore

File: movies.py (hunk 36 above)
Comment (CoitThomas) [inline movies.py]: ignore

File: movies_test.py [hunk 39]
```diff
@@ -0,0 +1,153 @@
...
+    bobs_actual_cash = movies.charge(bob)["cash"]
+    bobs_expected_cash = 95.0
+    assert bobs_actual_cash == bobs_expected_cash
+
+    jim = customers["Jim"]
+    assert movies.enough_money(bob)
+    jims_actual_cash = movies.charge(jim)["cash"]
+    jims_expected_cash = 5.0
+    assert jims_actual_cash == jims_expected_cash
+    ricci = customers["Ricci"]
```
Comment (CoitThomas) [inline movies_test.py]: give this some space too - it's testing something different

File: movies.py [hunk 40]
```diff
@@ -0,0 +1,80 @@
+"""Movie ticket process system"""
+
+MOVIE_PRICE = 5.0
+
+def enough_money(customer):
+    """Check if customer has adequate funds"""
+    balance = customer["cash"]
+    if balance < MOVIE_PRICE:
```
Comment (CoitThomas) [inline movies.py]: sorry, I should have seen this sooner, but if your logic is this simple, you can do some pretty sweet syntactic sugar in python: `return False if customer["cash"] < MOVIE_PRICE else True`

**Review by CoitThomas — APPROVED** [review]
🍄

File: README.md [hunk 41]
```diff
@@ -0,0 +1,17 @@
+# sywtltc_chapter_3_4
+Code challenge for So you want to learn to code 3.4
+
+Inherit some very interesting code from a former employee and reworked it to get it running and readable.
+
+purchase_tickets contains the subfunctions charge, dispense_ticket, and remove_seat
+refund_tickets contains add_funds, remove_ticket, and add_seat.
+
+It is only necessary to use purchase_tickets and refund_tickets. The subfunctions should not be necessary on their own.
```
Comment (austinwiltshire) [inline README.md]: Break things into two files - purchase_tickets and refund_tickets. 

You can create a third file that has your shared globals - so purchase_tickets and refund_tickets will both import something like movie_database.py that has the globals like your price in them. 

All functions that you don't want to be used on their own, prepend them with a _, so add_funds, since it shouldn't be used on its own, should be renamed _add_funds

Python doesn't do this well, but we're getting closer and closer to 'object orientation' and what you need for add_funds, etc... that aren't necessary on their own is called "encapsulation". Other languages have sections called 'public' and 'private' for this.

We'll simulate that here.

File: movies.py [hunk 42]
```diff
@@ -0,0 +1,77 @@
+"""Movie ticket process system"""
+
+MOVIE_PRICE = 5.0
+
+def enough_money(customer):
+    """Check if customer has adequate funds"""
+    return False if customer["cash"] < MOVIE_PRICE else True
```
Comment (austinwiltshire) [inline movies.py]: you can shorten this to simply

```
return customer["cash"] >= MOVIE_PRICE
```

File: movies.py [hunk 43]
```diff
@@ -0,0 +1,77 @@
...
+
+def purchase_ticket(movie, customer):
+    """Performs the steps required in purchasing a ticket and returns altered customer and movie"""
+    assert enough_money(customer)
+    customer = dispense_ticket(movie, charge(customer))
+    movie = remove_seat(movie)
+    return customer, movie
```
Comment (austinwiltshire) [inline movies.py]: i like this returning of the variables.

you could probably put them in the same order as you took them in though for consistency

File: movies.py [hunk 44]
```diff
@@ -0,0 +1,77 @@
...
+    return movie
+
+def print_movie_data(movie):
+    """Prints movie database informations"""
+    print("Movie Title:           ", movie["title"])
+    print("Seats Remaining:       ", movie["seats_available"])
+
+def print_customer_data(customer): #movies have brackets around them
+    """Prints customer database information"""
+    print("Movie Tickets:         ", ",  ".join(customer["movies"]))
```
Comment (austinwiltshire) [inline movies.py]: nice use of join

File: movies.py [hunk 45]
```diff
@@ -0,0 +1,77 @@
...
+    print("Account Balance:       ", customer["cash"])
+
+def print_all_movie_data(movie_dictionary):
+    """Print all movie data raw"""
+    for _, details in movie_dictionary.items():
```
Comment (austinwiltshire) [inline movies.py]: for dictionaries, you can iterate over just the values by using movie_dictionary.itervalues()

File: movies_test.py [hunk 46]
```diff
@@ -0,0 +1,160 @@
+"""Tests the functions of the code.py file"""
+
+import movies
+
+def customer_db():
```
Comment (austinwiltshire) [inline movies_test.py]: i like these test helpers.

File: movies_test.py [hunk 47]
```diff
@@ -0,0 +1,160 @@
...
+
+def test_charge():
+    """Tests charge function"""
+    customers = customer_db()
+
+    bob = customers["Bob"]
+    assert bob["cash"] == 100.0
+    assert movies.enough_money(bob)
+    bobs_actual_cash = movies.charge(bob)["cash"]
+    bobs_expected_cash = 95.0
```
Comment (austinwiltshire) [inline movies_test.py]: really good use of names to make the code readable

**Review by austinwiltshire — CHANGES_REQUESTED** [review]
few more structural and minor changes

File: test_refund_tickets_details.py [hunk 48]
```diff
@@ -0,0 +1,71 @@
...
+                 "cash" : 100.0},
+        "Jim" : {"movies": ["Xmen 8: The Xmennening"],
+                 "cash" : 10.0},
+        "Cary" : {"movies": ["Gigli: The Play: The Book: The movie", "The Bromance"],
+                  "cash" : 120.0},
+        "Ricci": {"movies": [],
+                  "cash" : 4.0}
+    }
+
+def movie_db():
```
Comment (austinwiltshire) [inline test_refund_tickets_details.py]: refactor out this function and the customer db function into a test_helpers.py module and import that into each of these test files

**Review by austinwiltshire — APPROVED** [review]
